bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs

```diff
@@ -0,0 +1,906 @@
+		}
+
+		for asset in &message.assets {
+			match asset {
+				EthereumAsset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					let asset: Asset = (token_location, *value).into();
+					assets.push(AssetTransfer::ReserveDeposit(asset));
+				},
+				EthereumAsset::ForeignTokenERC20 { token_id, value } => {
+					let asset_loc = ConvertAssetId::convert(&token_id)
+						.ok_or(ConvertMessageError::InvalidAsset)?;
+					let mut reanchored_asset_loc = asset_loc.clone();
+					reanchored_asset_loc
+						.reanchor(&GlobalAssetHubLocation::get(), &EthereumUniversalLocation::get())
+						.map_err(|_| ConvertMessageError::CannotReanchor)?;
+					let asset: Asset = (reanchored_asset_loc, *value).into();
+					assets.push(AssetTransfer::ReserveWithdraw(asset));
+				},
+			}
+		}
+
+		// Add SetTopic instruction if not already present as the last instruction
+		if !matches!(remote_xcm.0.last(), Some(SetTopic(_))) {
+			let topic = blake2_256(&(INBOUND_QUEUE_TOPIC_PREFIX, message.nonce).encode());
+			remote_xcm.0.push(SetTopic(topic));
+		}
+
+		let prepared_message = PreparedMessage {
+			origin: message.origin.clone(),
+			claimer,
+			assets,
+			remote_xcm,
+			execution_fee: execution_fee_asset,
+		};
+
+		Ok(prepared_message)
+	}
+
+	/// Get the bridge owner account ID from the current Ethereum network chain ID.
+	/// Returns an error if the network is not Ethereum.
+	fn bridge_owner() -> Result<[u8; 32], ConvertMessageError> {
+		let chain_id = match EthereumNetwork::get() {
+			NetworkId::Ethereum { chain_id } => chain_id,
+			_ => return Err(ConvertMessageError::InvalidNetwork),
+		};
+		Ok(EthereumLocationsConverterFor::<[u8; 32]>::from_chain_id(&chain_id))
+	}
+
+	/// Construct the remote XCM needed to create a new asset in the `ForeignAssets` pallet
+	/// on AssetHub. Polkadot is the only supported network at the moment.
+	fn make_create_asset_xcm(
+		token: &H160,
+		network: super::message::Network,
+		eth_value: u128,
+		bridge_owner: [u8; 32],
+		claimer: Location,
+	) -> Result<Xcm<()>, ConvertMessageError> {
+		let dot_asset = Location::new(1, Here);
+		let dot_fee: xcm::prelude::Asset = (dot_asset, CreateAssetDeposit::get()).into();
+
+		let eth_asset: xcm::prelude::Asset =
+			(Location::new(2, [GlobalConsensus(EthereumNetwork::get())]), eth_value).into();
+
+		let create_call_index: [u8; 2] = CreateAssetCall::get();
+
+		let asset_id = Location::new(
+			2,
+			[
+				GlobalConsensus(EthereumNetwork::get()),
+				AccountKey20 { network: None, key: (*token).into() },
+			],
+		);
+
+		// If the claimer is an AccountId32 on AH, use it to refund excess fees.
+		// Otherwise, use the bridge owner.
+		let claimer_account = match claimer.unpack() {
+			(0, [AccountId32 { id, .. }]) => *id,
+			_ => bridge_owner,
+		};
+
```

Comment:
Why force the claimer to be an `AccountId32 on AH`? Your API takes a location, so why not support any location?

Just use whatever the user set in the `claimer` - it can be any location (like Hydration for example), in which case the respective sov account is automatically used

(note that `claimer` here has already been populated by calling function with user input or `bridge_owner`)

```suggestion
```

Reply:
@acatangiu the claimer can be any location provided. The part of code you linked is specifically to refund excess fees after the Transact to create a foreign asset: https://github.com/paritytech/polkadot-sdk/blob/506279aef078d208e2b7107fb5c3bbe6c7d11eef/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L274 I understand the beneficiary for the `DepositAsset` needs to be an `AccountId32`, not a location. Please let me know if there is a better way.

Reply:
> the beneficiary for the DepositAsset needs to be an AccountId32, not a location

no, it can be any location [as the API shows](https://github.com/paritytech/polkadot-sdk/blob/007565a7225cba59d236ffc2fdaca2438a3665c1/polkadot/xcm/src/v5/mod.rs#L614)

any location will be converted to its sovereign account (an `AccountId32`)

Reply:
I see. Fixed in https://github.com/paritytech/polkadot-sdk/pull/7402/commits/2615c7d1f2a4c4503b14ec0ca1fae6fc4d893cc8.